ACR/Key Vault deploy relocation and moving to Flux extension (#71)
mspnp/aks-baseline-regulated: docs/deploy/11-validate-cluster-access-and-bootstrapping.md

# Validate cluster access and boostraping 

Now that [the AKS cluster](./10-aks-cluster.md) has been deployed, and your [bootstrapping images have passed through quarantine](./09-pre-cluster-stamp.md), the next step is to validate cluster access and boostraping. GitOps management solution was configure on your cluster, Flux in this case.

## Expected results

### Jump box access is validated

> While the following process likely would be handled via your deployment pipelines, we are going to use this opportunity to demonstrate cluster management access via Azure Bastion, and show that your cluster cannot be directly accessed locally.

Reviewer: ```suggestion
Because the cluster is private, your cluster cannot be directly accessed locally. You'll validate controlled jump box access for direct ops against the cluster if needed.
```

> ### AKS access validation

Reviewer: ```suggestion
### Bootstrapping results validation
```

> Your cluster was deployed with Azure Policy and Flux extension. You'll execute some commands to show how those add-ons are manifesting itself in your cluster. 

Reviewer: ```suggestion
Your cluster was deployed with Azure Policy and the Flux GitOps extension. You'll execute some commands to show how those are manifesting itself in your cluster.
```

## Steps

1. Connect to a jump box node via Azure Bastion.

>    If this is the first time you've used Azure Bastion, here is a detailed walk through of this process.

Reviewer: below (a ways) is a output section that still shows 1.22 for versions, can you please update that to 1.24 (or whatever version we're on here)

Author: I updated to the current cluster version, 1.23.12.

   ```output
   NAME                                  STATUS   ROLES   AGE   VERSION

>    aks-npinscope01-26621167-vmss000000   Ready    agent   20m   v1.23.12
>    aks-npinscope01-26621167-vmss000001   Ready    agent   20m   v1.23.12
>    aks-npooscope01-26621167-vmss000000   Ready    agent   20m   v1.23.12
>    aks-npooscope01-26621167-vmss000001   Ready    agent   20m   v1.23.12
>    aks-npsystem-26621167-vmss000000      Ready    agent   20m   v1.23.12
>    aks-npsystem-26621167-vmss000001      Ready    agent   20m   v1.23.12
>    aks-npsystem-26621167-vmss000002      Ready    agent   20m   v1.23.12

Reviewer: ```suggestion
   aks-npinscope01-26621167-vmss000000   Ready    agent   20m   v1.23.x
   aks-npinscope01-26621167-vmss000001   Ready    agent   20m   v1.23.x
   aks-npooscope01-26621167-vmss000000   Ready    agent   20m   v1.23.x
   aks-npooscope01-26621167-vmss000001   Ready    agent   20m   v1.23.x
   aks-npsystem-26621167-vmss000000      Ready    agent   20m   v1.23.x
   aks-npsystem-26621167-vmss000001      Ready    agent   20m   v1.23.x
   aks-npsystem-26621167-vmss000002      Ready    agent   20m   v1.23.x
```

   k8sazurevolumetypes                      21m
   ```

> 1. _From your Azure Bastion connection_, bootstrap Flux Validation.

Reviewer: ```suggestion
1. _From your Azure Bastion connection_, validate bootstrapping.
```

>    Validate that Flux has been bootstrapped.

Reviewer: ```suggestion
   Validate that the Flux extension has bootstrapped your cluster.
```

   ```output
   kubectl get namespaces

>    # You must see namespaces created by flux like a0005-i

Reviewer: ~~~suggestion
   ```bash
   kubectl get namespaces
   ```
   
   You should see namespaces created as described your manifest files, such as `a0005-i`.
   
   ```output
~~~

   NAME                        STATUS   AGE
   a0005-i                     Active   79m
   a0005-o                     Active   79m
   cluster-baseline-settings   Active   79m
   default                     Active   107m
   falco-system                Active   79m
   flux-system                 Active   87m
   gatekeeper-system           Active   107m
   ingress-nginx               Active   87m
   kube-node-lease             Active   107m
   kube-public                 Active   107m
   kube-system                 Active   107m
   ```

> There is a subnet allocated in this reference implementation called `snet-management-agents` specifically for your build agents to perform unattended, "last mile" deployment, and configuration needs for your cluster. There is no compute deployed to that subnet, but typically this is where you'd put in a VM Scale Set as a [GitHub Action Self-Hosted Runner](https://docs.github.com/actions/hosting-your-own-runners/about-self-hosted-runners) or [Azure DevOps Self-Hosted Agent Pool](https://learn.microsoft.com/azure/devops/pipelines/agents/scale-set-agents?view=azure-devops). This compute should be a hardened, minimal installation set, and monitored. Just like a jump box, this compute will span two distinct security zones; in this case, unattended, externally managed GitHub Workflow definitions and your cluster.

Reviewer: ```suggestion
## Configuration needs outside of GitOps

There is a subnet allocated in this reference implementation called `snet-management-agents` specifically for your build agents to perform unattended, "last mile" deployment, and configuration needs for your cluster. There is no compute deployed to that subnet, but typically this is where you'd put in a Virtual Machine Scale Set as a [GitHub Action Self-Hosted Runner](https://docs.github.com/actions/hosting-your-own-runners/about-self-hosted-runners) or [Azure DevOps Self-Hosted Agent Pool](https://learn.microsoft.com/azure/devops/pipelines/agents/scale-set-agents?view=azure-devops). This compute should be a hardened, minimal installation set, and monitored. Just like a jump box, this compute will span two distinct security zones; in this case, unattended, externally managed GitHub Workflow definitions and your cluster.
```

> In summary, there should be a stated objective in your team that routine processes (especially bootstrapping new clusters), are never performed by a human via direct access tools `kubectl` -- _Zero-kubectl starting from Day 0_. Any ad-hoc human interaction with the cluster introduces risk and audit trail concerns. Obviously, live-site issues will require humans to perform out-of-the-norm practices to triage and address critical issues. Reserve your usage of direct access tooling like this for those times.

Reviewer: Below in the "GitOps configuration" section (can't leave a comment there, sorry).

* remove "Built-in [bootstrapping support](https://fluxcd.io/docs/installation/#bootstrap)"
* remove last three items (all start with "Flux's ")
* add "Additional flux controllers like Helm. These are explicitly disabled to reduce surface area on the extension in the regulated cluster.  Enable only those controllers you need from within the `mcFlux_extension` resource in your ARM template."

Also add

In summary, there should be a stated objective in your team that routine processes (especially bootstrapping new clusters), are never performed by a human via direct access tools `kubectl` -- _Zero-kubectl starting from Day 0_. Any ad-hoc human interaction with the cluster introduces risk and audit trail concerns. Obviously, live-site issues will require humans to perform out-of-the-norm practices to triage and address critical issues. Reserve your usage of direct access tooling like this for those times.

## GitOps configuration

The GitOps implementation in this reference architecture is _intentionally simplistic_. Flux is configured to simply monitor manifests in ALL namespaces. It doesn't account for concepts like:

* [Multi-tenancy](https://github.com/fluxcd/flux2-multi-tenancy)
* [Private GitHub repos](https://fluxcd.io/docs/components/source/gitrepositories/#ssh-authentication)
* Kustomization [under/overlays](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/kustomization/#bases-and-overlays)

> 
> Additional flux controllers like Helm. These are explicitly disabled to reduce surface area on the extension in the regulated cluster. Enable only those controllers you need from within the mcFlux_extension resource in your ARM template.

Reviewer: ```suggestion
* Additional Flux controllers like Helm. These additional controllers are explicitly disabled to reduce surface area on the extension in the regulated cluster. Enable only those controllers you need from within the `mcFlux_extension` resource in your ARM template.
```
